// nmea-simple/dist/tests/GLLtest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("should");
var index_1 = require("../index");
describe("GLL", function () {
    it("parser", function () {
        var packet = index_1.parseNmeaSentence("$GPGLL,6005.068,N,02332.341,E,095601,A,D*42");
        packet.should.have.property("sentenceId", "GLL");
        packet.should.have.property("sentenceName", "Geographic position - latitude and longitude");
        packet.should.have.property("talkerId", "GP");
        packet.should.have.property("latitude", 60.084466666666664);
        packet.should.have.property("longitude", 23.539016666666665);
        var timestamp = new Date();
        timestamp.setUTCHours(9);
        timestamp.setUTCMinutes(56);
        timestamp.setUTCSeconds(1);
        timestamp.setUTCMilliseconds(0);
        packet.should.have.property("time", timestamp);
        packet.should.have.property("status", "valid");
        packet.should.have.property("faaMode", "D");
    });
    it("encoder", function () {
        var timestamp = new Date();
        timestamp.setUTCHours(21);
        timestamp.setUTCMinutes(17);
        timestamp.setUTCSeconds(22);
        timestamp.setUTCMilliseconds(0);
        var sentence = index_1.encodeNmeaPacket({
            sentenceId: "GLL",
            latitude: 60.084333,
            longitude: 23.539,
            time: timestamp,
            status: "valid",
            faaMode: "D"
        }, "II");
        sentence.should.equal("$IIGLL,6005.059980,N,02332.340000,E,211722,A,D*59");
    });
});
